doc/man/man1/hash.1
.\" Automatically generated by Pandoc 1.16.0.2
.\"
.TH "HASH" "1" "February 29, 2016" "Virgil Security CLI (2.0.0)" "Virgil"
.hy
.SH NAME
.PP
\f[B]hash\f[] \[en] derives the obfuscated data from incoming parameters
using PBKDF function.
.SH SYNOPSIS
.IP
.nf
\f[C]
\ \ \ \ virgil\ hash\ \ [\-i\ <file>]\ [\-o\ <file>]\ \-s\ <file>\ [\-a\ <sha1|sha224|sha256|sha384
\ \ \ \ \ \ \ \ \ |sha512>]\ [\-c\ <int>]\ [\-V]\ [\-\-]\ [\-\-version]\ [\-h]
\f[]
.fi
.SH DESCRIPTION
.PP
Derives the obfuscated data from incoming parameters using PBKDF
function.
.SH OPTIONS
.IP
.nf
\f[C]
\ \ \ \ \-i\ <file>,\ \ \-\-in\ <file>
\ \ \ \ \ The\ string\ value\ to\ be\ hashed.\ If\ omitted,\ stdout\ is\ used.

\ \ \ \ \-o\ <file>,\ \ \-\-out\ <file>
\ \ \ \ \ Obfuscated\ data.\ If\ omitted,\ stdout\ is\ used.

\ \ \ \ \-s\ <string>,\ \ \-\-salt\ <file>
\ \ \ \ \ (required)\ \ The\ hash\ salt.

\ \ \ \ \-a\ <sha1|sha224|sha256|sha384|sha512>,\ \ \-\-algorithm\ <sha1|sha224|sha256
\ \ \ \ \ \ |sha384|sha512>
\ \ \ \ \ Generate\ hash\ with\ oneof\ the\ following\ positions:

\ \ \ \ \ \ \ \ *\ sha1\ \-\ \ \ secure\ Hash\ Algorithm\ 1;

\ \ \ \ \ \ \ \ *\ sha224\ \-\ hash\ algorithm;

\ \ \ \ \ \ \ \ *\ sha256\ \-\ hash\ algorithm;

\ \ \ \ \ \ \ \ *\ sha384\ \-\ hash\ algorithm(default);

\ \ \ \ \ \ \ \ *\ sha512\ \-\ hash\ algorithm;


\ \ \ \ \-c\ <int>,\ \ \-\-iterations\ <int>
\ \ \ \ \ The\ count\ of\ iterations.\ Default\ \-\ 2048

\ \ \ \ \-V,\ \ \-\-VERBOSE
\ \ \ \ \ Show\ detailed\ information

\ \ \ \ \-\-,\ \ \-\-ignore_rest
\ \ \ \ \ Ignores\ the\ rest\ of\ the\ labeled\ arguments\ following\ this\ flag.

\ \ \ \ \-\-version
\ \ \ \ \ Displays\ version\ information\ and\ exits.

\ \ \ \ \-h,\ \ \-\-help
\ \ \ \ \ Displays\ usage\ information\ and\ exits.
\f[]
.fi
.SH EXAMPLES
.IP "1." 3
Generate hash (alg \- sha384, iterations \- 2048 default):
.RS 4
.IP
.nf
\f[C]
virgil\ hash\ \-i\ data.txt\ \-o\ obfuscated_data.txt\ \-s\ data_salt.txt
\f[]
.fi
.RE
.IP "2." 3
Generate hash sha512 and count of iterations \- 4096:
.RS 4
.IP
.nf
\f[C]
virgil\ hash\ \-i\ data.txt\ \-o\ obfuscated_data.txt\ \-s\ data_salt.txt\ \-a\ sha512\ \-c\ 4096
\f[]
.fi
.RE
.SH SEE ALSO
.PP
virgil(1)
.PD 0
.P
.PD
card\-create\-private(1)
.SH AUTHORS
(c) Virgil Security Inc.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version v2.0.0-rc1 released 2016-05-16

Changes:
  * Use virgil-crypto lib 1.6.0
  * Use virgil-sdk-cpp lib 3.2.0
  * Add config file


= Version 0.3.4 released 2015-08-27

Bugfix:
  * Update underlying virgil crypto library to fixed version 1.0.1.

= Version 0.3.3 released 2015-08-18

Bugfix:
  * Embed MSVC redistributable package installation to the CLI package.

Changes:
  * Add executable folder to the PATH environment variable silently on Windows.

= Version 0.3.2 released 2015-08-17

Bugfix:
  * Fix RSA and EC keys generation.

= Version 0.3.1 released 2015-08-17

Bugfix:
  * Add missing win32 runtime libraries.

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/base/hash.cxx
/**
 * Copyright (C) 2015 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <iostream>
#include <stdexcept>
#include <string>
#include <fstream>
#include <iterator>

#include <tclap/CmdLine.h>

#include <virgil/crypto/foundation/VirgilPBKDF.h>

#include <virgil/sdk/util/obfuscator.h>

#include <cli/version.h>
#include <cli/pair.h>
#include <cli/util.h>

namespace vcrypto = virgil::crypto;
namespace vsdk = virgil::sdk;
namespace vcli = virgil::cli;

#ifdef SPLIT_CLI
#define MAIN main
#else
#define MAIN hash_main
#endif

static vcrypto::foundation::VirgilPBKDF::Hash hash_alg(const std::string& param);

int MAIN(int argc, char** argv) {
    try {
        std::string description = "Derives the obfuscated data from incoming parameters using PBKDF function.\n\n";

        std::vector<std::string> examples;
        examples.push_back("Generate hash (alg - sha384, iterations - 2048 default):\n"
                           "virgil hash -i data.txt -o obfuscated_data.txt -s data_salt.txt\n\n");

        examples.push_back("Generate hash sha512 and count of iterations - 4096:\n"
                           "virgil hash -i data.txt -o obfuscated_data.txt -s data_salt.txt -a sha512 -c 4096\n\n");

        std::string descriptionMessage = vcli::getDescriptionMessage(description, examples);

        // Parse arguments.
        TCLAP::CmdLine cmd(descriptionMessage, ' ', virgil::cli_version());

        TCLAP::ValueArg<std::string> inArg("i", "in", "The string value to be hashed. If omitted, stdout is used.",
                                           false, "", "file");

        TCLAP::ValueArg<std::string> outArg("o", "out", "Obfuscated data. If omitted, stdout is used.", false, "",
                                            "file");

        std::vector<std::string> alg;
        alg.push_back("sha1");
        alg.push_back("sha224");
        alg.push_back("sha256");
        alg.push_back("sha384");
        alg.push_back("sha512");
        TCLAP::ValuesConstraint<std::string> allowedAlg(alg);

        TCLAP::ValueArg<std::string> saltArg("s", "salt", "The hash salt.", true, "", "file");

        TCLAP::ValueArg<std::string> algorithmArg("a", "algorithm", "Generate hash with one"
                                                                    "of the following positions:\n"
                                                                    "\t* sha1 -   secure Hash Algorithm 1;\n"
                                                                    "\t* sha224 - hash algorithm;\n"
                                                                    "\t* sha256 - hash algorithm;\n"
                                                                    "\t* sha384 - hash algorithm(default);\n"
                                                                    "\t* sha512 - hash algorithm;\n",
                                                  false, "sha384", &allowedAlg);

        TCLAP::ValueArg<int> iterationsArg("c", "iterations", "The count of iterations. Default - 2048", false, 2048,
                                           "int");

        TCLAP::SwitchArg verboseArg("V", "VERBOSE", "Show detailed information", false);

        cmd.add(verboseArg);
        cmd.add(iterationsArg);
        cmd.add(algorithmArg);
        cmd.add(saltArg);
        cmd.add(outArg);
        cmd.add(inArg);

        cmd.parse(argc, argv);

        if (verboseArg.isSet()) {
            std::cout << "Generating hash.." << std::endl;
        }

        std::string pathToSaltFile = saltArg.getValue();
        std::ifstream inSaltFile(pathToSaltFile, std::ios::in | std::ios::binary);
        if (!inSaltFile) {
            throw std::invalid_argument("can not read salt file by path: " + pathToSaltFile);
        }

        std::string salt((std::istreambuf_iterator<char>(inSaltFile)), std::istreambuf_iterator<char>());

        std::string value = vcli::readInput(inArg.getValue());
        auto sequenceBase64 = vsdk::util::obfuscate(value, saltArg.getValue(), hash_alg(algorithmArg.getValue()),
                                                    iterationsArg.getValue());
        vcli::writeOutput(outArg.getValue(), sequenceBase64);
        if (verboseArg.isSet()) {
            std::cout << "The hash generated" << std::endl;
        }

    } catch (TCLAP::ArgException& exception) {
        std::cerr << "hash. Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "hash. Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

vcrypto::foundation::VirgilPBKDF::Hash hash_alg(const std::string& param) {
    std::map<std::string, vcrypto::foundation::VirgilPBKDF::Hash> hashAlg;
    hashAlg["sha1"] = vcrypto::foundation::VirgilPBKDF::Hash::Hash_SHA1;
    hashAlg["sha224"] = vcrypto::foundation::VirgilPBKDF::Hash::Hash_SHA224;
    hashAlg["sha256"] = vcrypto::foundation::VirgilPBKDF::Hash::Hash_SHA256;
    hashAlg["sha384"] = vcrypto::foundation::VirgilPBKDF::Hash::Hash_SHA384;
    hashAlg["sha512"] = vcrypto::foundation::VirgilPBKDF::Hash::Hash_SHA512;

    auto group = hashAlg.find(param);
    if (group != hashAlg.end()) {
        return group->second;
    }

    return vcrypto::foundation::VirgilPBKDF::Hash::Hash_SHA384;
}
